### PR TITLE
Support XINFO STREAM FULL

### PR DESCRIPTION
**Pull request by zeekling**
support XINFO STREAM key FULL  [COUNT count] like:
```bash
xinfo stream xadd-stream1 full count 10
```

### CODE REVIEW

**Review by yangbodong22011 — COMMENTED** [review]
The current code is good for me, but the corresponding commands need to be added to the following classes:
- Pipeline.java
- TransactionBase.java
- MultiNodePipelineBase.java

**Comment by zeekling** [comment]
> The current code is good for me, but the corresponding commands need to be added to the following classes:
> 
> * Pipeline.java
> * TransactionBase.java
> * MultiNodePipelineBase.java



> The current code is good for me, but the corresponding commands need to be added to the following classes:
> 
> * Pipeline.java
> * TransactionBase.java
> * MultiNodePipelineBase.java

done

**Comment by yangbodong22011** [comment]
Seems the test `ci/circleci: build-linux-8-jdk Expected — Waiting for status to be reported` does not pass 
@sazzad16 Do you know or can check the reason?
image

**Comment by sazzad16** [comment]
@chayim PTAL ^^^

File: src/test/java/redis/clients/jedis/commands/jedis/StreamsCommandsTest.java [hunk 1]
```diff
@@ -746,6 +746,14 @@ public void xinfo() throws InterruptedException {
     assertEquals(2, manyGroupsInfo.size());
     assertEquals(2, manyConsumersInfo.size());
 
+    StreamFullInfo streamInfoFull = jedis.xinfoStreamFull(STREAM_NAME);
+
+    assertEquals(2, streamInfoFull.getEntries().size());
+    assertEquals(2, streamInfoFull.getGroups().size());
+    assertEquals(2, streamInfoFull.getLength());
+    assertEquals(1, streamInfoFull.getRadixTreeKeys());
+    assertEquals(2, streamInfoFull.getRadixTreeNodes());
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/commands/jedis/StreamsCommandsTest.java:755]: Can we make tests better?

File: src/main/java/redis/clients/jedis/BuilderFactory.java [hunk 2]
```diff
@@ -1109,6 +1112,128 @@ public String toString() {
     }
   };
 
+  public static final Builder<List<ConsumerFullInfo>> STREAM_CONSUMER_FULL_INFO_LIST = new Builder<List<ConsumerFullInfo>>() {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BuilderFactory.java:1115]:
```suggestion
  private static final Builder<List<ConsumerFullInfo>> STREAM_CONSUMER_FULL_INFO_LIST = new Builder<List<ConsumerFullInfo>>() {
```

File: src/main/java/redis/clients/jedis/BuilderFactory.java [hunk 3]
```diff
@@ -1109,6 +1112,128 @@ public String toString() {
...
+      return list;
+    }
+
+    @Override
+    public String toString() {
+      return "List<ConsumerInfo>";
+    }
+  };
+
+  public static final Builder<List<StreamGroupFullInfo>> STREAM_GROUP_FULL_INFO_LIST = new Builder<List<StreamGroupFullInfo>>() {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/BuilderFactory.java:1155]:
```suggestion
  private static final Builder<List<StreamGroupFullInfo>> STREAM_GROUP_FULL_INFO_LIST = new Builder<List<StreamGroupFullInfo>>() {
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]
`...FullInfo` classes should be put into `resps` package.

File: src/main/java/redis/clients/jedis/params/ConsumerFullInfo.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class holds information about a stream consumer with command <code>xinfo stream mystream full<code/>.
+ * They can be access via getters. For future purpose there is also {@link #getConsumerInfo()} method that
+ * returns a generic {@code Map} - in case where more info is returned from the server.
+ */
+public class ConsumerFullInfo implements Serializable {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/params/ConsumerFullInfo.java:12]: ConsumerFullInfo should also be put into `resps` package.

**Review by sazzad16 — CHANGES_REQUESTED** [review]
@zeekling Please address the change requests.

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 5]
```diff
@@ -4466,6 +4466,31 @@ public Object xinfoStream(byte[] key) {
...
+  @Override
+  public StreamFullInfo xinfoStreamFull(byte[] key, int count) {
+    checkIsInMultiOrPipeline();
+    return connection.executeCommand(commandObjects.xinfoStreamFull(key, count));
+  }
+
+  @Override
+  public StreamFullInfo xinfoStreamFull(String key) {
+    checkIsInMultiOrPipeline();
+    return connection.executeCommand(commandObjects.xinfoStreamFull(key, 10));
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java:4485]: Don't add `10`.

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 6]
```diff
@@ -4466,6 +4466,31 @@ public Object xinfoStream(byte[] key) {
     return connection.executeCommand(commandObjects.xinfoStream(key));
   }
 
+  @Override
+  public StreamFullInfo xinfoStreamFull(byte[] key) {
+    checkIsInMultiOrPipeline();
+    return connection.executeCommand(commandObjects.xinfoStreamFull(key, 10));
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java:4472]: Don't add `10`.

File: src/main/java/redis/clients/jedis/UnifiedJedis.java [hunk 7]
```diff
@@ -2465,6 +2465,26 @@ public StreamInfo xinfoStream(String key) {
     return executeCommand(commandObjects.xinfoStream(key));
   }
 
+  @Override
+  public StreamFullInfo xinfoStreamFull(byte[] key) {
+    return executeCommand(commandObjects.xinfoStreamFull(key, 10));
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/UnifiedJedis.java:2470]: Don't add `10`.

File: src/main/java/redis/clients/jedis/UnifiedJedis.java [hunk 8]
```diff
@@ -2465,6 +2465,26 @@ public StreamInfo xinfoStream(String key) {
...
+  }
+
+  @Override
+  public StreamFullInfo xinfoStreamFull(byte[] key, int count) {
+    return executeCommand(commandObjects.xinfoStreamFull(key, count));
+  }
+
+  @Override
+  public StreamFullInfo xinfoStreamFull(String key) {
+    return executeCommand(commandObjects.xinfoStreamFull(key, 10));
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/UnifiedJedis.java:2480]: Don't add `10`.

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/params/ConsumerFullInfo.java (hunk 4 above)
Comment (zeekling) [reply src/main/java/redis/clients/jedis/params/ConsumerFullInfo.java]: moved

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 5 above)
Comment (zeekling) [reply src/main/java/redis/clients/jedis/Jedis.java:4484]: deleted

File: src/test/java/redis/clients/jedis/commands/jedis/StreamsCommandsTest.java [hunk 9]
```diff
@@ -752,6 +752,20 @@ public void xinfo() throws InterruptedException {
     assertEquals(2, manyGroupsInfo.size());
     assertEquals(2, manyConsumersInfo.size());
 
+    StreamFullInfo streamInfoFull = jedis.xinfoStreamFull(STREAM_NAME);
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/commands/jedis/StreamsCommandsTest.java:755]: Please add one test with count.

File: src/main/java/redis/clients/jedis/resps/StreamGroupFullInfo.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    private final List<String> pending;
+    private final Long pelCount;
+    private final StreamEntryID lastDeliveredId;
+    private final Map<String, Object> groupFullInfo;
+
+    /**
+     * @param map contains key-value pairs with group info
+     */
+    @SuppressWarnings("unchecked")
+    public StreamGroupFullInfo(Map<String, Object> map) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/resps/StreamGroupFullInfo.java:33]: Please format with 2 spaces as tabs.

File: src/main/java/redis/clients/jedis/resps/StreamFullInfo.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+    private final List<StreamGroupFullInfo> groups;
+    private final StreamEntryID lastGeneratedId;
+    private final List<StreamEntry> entries;
+    private final Map<String, Object> streamFullInfo;
+
+    /**
+     * @param map contains key-value pairs with stream info
+     */
+    @SuppressWarnings("unchecked")
+    public StreamFullInfo(Map<String, Object> map) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/resps/StreamFullInfo.java:35]: Please format with 2 spaces as tabs.

File: src/main/java/redis/clients/jedis/resps/ConsumerFullInfo.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    public static final String PENDING = "pending";
+
+    private final String name;
+    private final Long seenTime;
+    private final Long pelCount;
+    private final List<Long> pending;
+    private final Map<String, Object> consumerInfo;
+
+    @SuppressWarnings("unchecked")
+    public ConsumerFullInfo(Map<String, Object> map){
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/resps/ConsumerFullInfo.java:26]: Please format with 2 spaces as tabs.

File: src/main/java/redis/clients/jedis/commands/StreamPipelineBinaryCommands.java [hunk 13]
```diff
@@ -60,6 +61,14 @@
 
   Response<Object> xinfoStream(byte[] key);
 
+  Response<StreamFullInfo> xinfoStreamFull(byte[] key);
+
+  Response<StreamFullInfo> xinfoStreamFull(byte[] key, int count);
+
+  Response<StreamFullInfo> xinfoStreamFull(String key);
+
+  Response<StreamFullInfo> xinfoStreamFull(String key, int count);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/commands/StreamPipelineBinaryCommands.java:70]: Move these in StreamPipelineCommands.

File: src/main/java/redis/clients/jedis/commands/StreamBinaryCommands.java [hunk 14]
```diff
@@ -57,6 +58,14 @@
 
   Object xinfoStream(byte[] key);
 
+  StreamFullInfo xinfoStreamFull(byte[] key);
+
+  StreamFullInfo xinfoStreamFull(byte[] key, int count);
+
+  StreamFullInfo xinfoStreamFull(String key);
+
+  StreamFullInfo xinfoStreamFull(String key, int count);
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/commands/StreamBinaryCommands.java:67]: Move these in StreamCommands.

File: src/main/java/redis/clients/jedis/UnifiedJedis.java (hunk 8 above)
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/UnifiedJedis.java:2480]:
```suggestion
    return executeCommand(commandObjects.xinfoStreamFull(key));
```

File: src/main/java/redis/clients/jedis/UnifiedJedis.java (hunk 7 above)
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/UnifiedJedis.java:2470]:
```suggestion
    return executeCommand(commandObjects.xinfoStreamFull(key));
```

File: src/main/java/redis/clients/jedis/TransactionBase.java [hunk 15]
```diff
@@ -1402,6 +1402,27 @@ public String discard() {
     return appendCommand(commandObjects.xinfoStream(key));
   }
 
+  @Override
+  public Response<StreamFullInfo> xinfoStreamFull(byte[] key) {
+    return appendCommand(commandObjects.xinfoStreamFull(key, 10));
+
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/TransactionBase.java:1408]:
```suggestion
    return appendCommand(commandObjects.xinfoStreamFull(key));
```

File: src/main/java/redis/clients/jedis/Pipeline.java [hunk 16]
```diff
@@ -2584,6 +2584,27 @@ public final boolean hasPipelinedResponse() {
     return appendCommand(commandObjects.xinfoStream(key));
   }
 
+  @Override
+  public Response<StreamFullInfo> xinfoStreamFull(byte[] key) {
+    return appendCommand(commandObjects.xinfoStreamFull(key, 10));
+
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Pipeline.java:2590]:
```suggestion
    return appendCommand(commandObjects.xinfoStreamFull(key));
```

File: src/main/java/redis/clients/jedis/MultiNodePipelineBase.java [hunk 17]
```diff
@@ -1348,6 +1348,27 @@ public final void sync() {
     return appendCommand(commandObjects.xinfoStream(key));
   }
 
+  @Override
+  public Response<StreamFullInfo> xinfoStreamFull(byte[] key) {
+    return appendCommand(commandObjects.xinfoStreamFull(key, 10));
+
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/MultiNodePipelineBase.java:1354]:
```suggestion
    return appendCommand(commandObjects.xinfoStreamFull(key));
```

File: src/main/java/redis/clients/jedis/Jedis.java [hunk 18]
```diff
@@ -4466,6 +4466,30 @@ public Object xinfoStream(byte[] key) {
...
+  @Override
+  public StreamFullInfo xinfoStreamFull(byte[] key, int count) {
+    checkIsInMultiOrPipeline();
+    return connection.executeCommand(commandObjects.xinfoStreamFull(key, count));
+  }
+
+  @Override
+  public StreamFullInfo xinfoStreamFull(String key) {
+    checkIsInMultiOrPipeline();
+    return connection.executeCommand(commandObjects.xinfoStreamFull(key, 10));
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/Jedis.java:4484]:
```suggestion
    return connection.executeCommand(commandObjects.xinfoStreamFull(key));
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/Jedis.java (hunk 18 above)
Comment (zeekling) [reply src/main/java/redis/clients/jedis/Jedis.java]: done

File: src/main/java/redis/clients/jedis/BuilderFactory.java [hunk 19]
```diff
@@ -1109,6 +1112,128 @@ public String toString() {
...
+        Iterator<Object> consumerInfoIterator = consumerInfoList.iterator();
+        ConsumerFullInfo consumerInfo = new ConsumerFullInfo(createMapFromDecodingFunctions(consumerInfoIterator, mappingFunctions));
+        list.add(consumerInfo);
+      }
+      return list;
+    }
+
+    @Override
+    public String toString() {
+      return "List<ConsumerInfo>";
```
Comment (dengliming) [inline src/main/java/redis/clients/jedis/BuilderFactory.java]:
```suggestion
      return "List<ConsumerFullInfo>";
```

File: src/main/java/redis/clients/jedis/resps/ConsumerFullInfo.java [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class holds information about a stream consumer with command <code>xinfo stream mystream full<code/>.
+ * They can be access via getters. For future purpose there is also {@link #getConsumerInfo()} method that
+ * returns a generic {@code Map} - in case where more info is returned from the server.
+ */
+public class ConsumerFullInfo implements Serializable {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/resps/ConsumerFullInfo.java:12]: Sorry for saying this late. But `StreamConsumerFullInfo` may be a better name here.

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/resps/ConsumerFullInfo.java (hunk 20 above)
Comment (zeekling) [reply src/main/java/redis/clients/jedis/resps/ConsumerFullInfo.java]: done

File: src/main/java/redis/clients/jedis/BuilderFactory.java (hunk 19 above)
Comment (zeekling) [reply src/main/java/redis/clients/jedis/BuilderFactory.java]: done

**Review by sazzad16 — APPROVED** [review]

**Pull request merged**